src/sinks/gcp_pubsub.rs

```diff
@@ -0,0 +1,294 @@
+    InvalidCredentials { source: GOErr },
+    #[snafu(display("Invalid RSA key in GCP credentials"))]
+    InvalidRsaKey { source: GOErr },
+    #[snafu(display("Failed to get OAuth token"))]
+    GetTokenFailed { source: GOErr },
+}
+
+inventory::submit! {
+    SinkDescription::new::<PubsubConfig>("gcp_pubsub")
+}
+
+lazy_static! {
+    static ref REQUEST_DEFAULTS: TowerRequestConfig = TowerRequestConfig {
+        ..Default::default()
+    };
+}
+
+#[typetag::serde(name = "gcp_pubsub")]
+impl SinkConfig for PubsubConfig {
+    fn build(&self, acker: Acker) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        if self.api_key.is_none() && self.credentials_path.is_none() {
+            return Err(BuildError::MissingAuth.into());
+        }
+
+        let creds = match self.credentials_path.as_ref() {
+            Some(path) => Some(PubsubCreds::new(path)?),
+            None => None,
+        };
+
+        let sink = self.service(acker, &creds)?;
+        let healthcheck = self.healthcheck(&creds)?;
+
+        Ok((sink, healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "gcp_pubsub"
+    }
+}
+
+impl PubsubConfig {
+    fn service(
+        &self,
+        acker: Acker,
+        creds: &Option<PubsubCreds>,
+    ) -> crate::Result<super::RouterSink> {
+        let batch = self.batch.unwrap_or(bytesize::mib(10u64), 1);
+        let request = self.request.unwrap_with(&REQUEST_DEFAULTS);
+
+        let uri = self.uri(":publish")?;
+        let tls_settings = TlsSettings::from_options(&self.tls)?;
+        let creds = creds.clone();
+
+        let http_service =
+            HttpService::builder()
+                .tls_settings(tls_settings)
+                .build(move |logs: Vec<u8>| {
+                    let mut builder = hyper::Request::builder();
+                    builder.method(Method::POST);
+                    builder.uri(uri.clone());
+                    builder.header("Content-Type", "application/x-json");
+
+                    let mut request = builder.body(make_body(logs)).unwrap();
+                    if let Some(creds) = creds.as_ref() {
+                        creds.apply(&mut request);
+                    }
+
+                    request
+                });
+
+        let sink = request
+            .batch_sink(HttpRetryLogic, http_service, acker)
+            .batched_with_min(Buffer::new(false), &batch)
+            .with_flat_map(|event| iter_ok(Some(encode_event(event))));
+
+        Ok(Box::new(sink))
+    }
+
+    fn healthcheck(&self, creds: &Option<PubsubCreds>) -> crate::Result<super::Healthcheck> {
+        let uri = self.uri("")?;
+        let mut request = Request::get(uri).body(Body::empty()).unwrap();
+        if let Some(creds) = creds.as_ref() {
+            creds.apply(&mut request);
+        }
+
+        let https = HttpsConnector::new(4).expect("TLS initialization failed");
```

Comment:
```suggestion
        let https = HttpsConnector::new(1).expect("TLS initialization failed");
```
probably don't need 4 threads for the dns resolver here :D

Reply:
Heh, of course not. Rather excessive. Fixed.